Update stopping test

frapac opened the pull request:
- Stop SDDP with the number of iterations
- Estimate the gap between the real solution and SDDP with Monte-Carlo
- fix some errors in code

frapac committed:
[DEV] Improve efficiency of backward_pass

+ vectorize computation of esperancy in backward_pass
+ add time of execution in stdout
+ add ! to name of backward_pass

frapac committed:
[DEV] Change stopping criterion in SDDP

- remove stopping criterion based on estimation of upper-boud
- update parameter of forwardPass to remove redundancy
- compute more accurate estimation of upper-bound at the end of the algorithm

frapac committed:
[FIX] Coordinate indexing between different arrays in forward_pass

frapac committed:
[FIX] Update calculation of initial cost in backward_pass

frapac committed:
[UPD] Test SDDP algorithm in runtests.jl with a simple example

+ update dams examples

Henri-Gerard commented on test/runtests.jl:
```diff
@@ -44,7 +43,62 @@ facts("Utility functions") do
+    # number of aleas:
+    n_aleas = 5
+    # number of stages:
+    n_stages = 2
```
Instead of defining a number of stages, isn't it better to define a number of time step to avoid confusion between state and control ? For example, when there is just a control, a two stage problem is a problem with two decision. But if we add a stock and if we consider the final stock, the number of times steps is 3

frapac replied:
You are right

leclere commented on src/utility.jl:
```diff
@@ -44,7 +44,7 @@ end
 """
-Estimate the upper bound with the Monte-Carlo error
+Estimate the upper bound with a distribution of costs
 
```
Give detail on the choosen error : 
the expectation of the costs have 97.5% chance of being under the upper bound.  

I would like to see the 1.96 not being so hard-coded, at some point we might want to
allow the user to change this parameter in one place (a global parameter in the package ?).

For further reference if you want P(E(cost) < upb )>= p, you need to use
sqrt(2) \* erfinv(2 \* p - 1.0) to compute the quantile.

leclere commented on src/forwardBackwardIterations.jl:
```diff
@@ -241,32 +229,35 @@ Return:
     T = model.stageNumber
 
     # Estimation of initial cost:
     V0 = 0.
```
I don't understand what this V0 is supposed to be ?

leclere commented on src/SDDPoptimize.jl:
```diff
@@ -208,42 +209,100 @@ function optimize(model::SPModel,
 
-        V0 = backward_pass(model,
+        # Backward pass
+        V0 = backward_pass!(model,
```
I don't understand this V0. It should be a lower bound (called lb for example), obtained by the value of the polyhedral function V0^k at x0. I do not see it computed this way in backward_pass!

leclere commented on src/SDDPoptimize.jl:
```diff
@@ -208,42 +209,100 @@ function optimize(model::SPModel,
+        upb = estimate_upper_bound(model, param, V, problems)
+        println("Estimation of upper-bound: ", upb,
+                "\tV0: ", V0,
+                "\t Gap (\%): ", (V0-upb)/V0)
```
At the moment it is "Gap  < (V0-upb)/V0) with prob. > 97.5\%"

leclere commented on src/SDDPoptimize.jl:
```diff
@@ -208,42 +209,100 @@ function optimize(model::SPModel,
+    if display
+        println("Estimate upper-bound with Monte-Carlo ...")
+        upb = estimate_upper_bound(model, param, V, problems)
+        println("Estimation of upper-bound: ", upb,
```
I feel that we should rework the presentation a bit. I would like to have the following display : 
- estimation of cost of the solution (fiability 95%) : mean(costs) +/- 1.96 stdev(costs)/sqrt(size(costs))
- exact lower bound : V0
-  "Gap < (V0-upb)/V0) with prob. > 97.5\%"

frapac committed:
[UPD] Rescope previous changes to update PR#17

- remove deprecated add_constraints_with_cut in forwardBackwardIterations.jl
- rename optimize as solve_SDDP and move this function at top of SDDPoptimize.jl
- update examples and test with new name of solve_SDDP
- update some docstrings
- add a function to compute estimated value of bellman functions at a given point

frapac committed:
[UPD] Add a more generic way to define terminal cost

Henri-Gerard commented on examples/damsvalley.jl:
```diff
@@ -170,7 +170,7 @@ end
     model, params = init_problem()
 
-    V, pbs = optimize(model, params, display)
+    V, pbs = solve_SDDP(model, params, display)
```
We have to set a reunion to brainstorm about the naming. Glad you bring the subject with this modification

Henri-Gerard commented on src/SDDPoptimize.jl:
```diff
@@ -169,81 +335,41 @@ end
+- t (Int64)
+    Time t where to solve bellman value
 
+- Vt (Polyhedral function)
```
For example, for the naming, is it a good idea to call the approximation of Value function by Vt ?
